### PR TITLE
feat(theme): Add high contrast themes

### PR DESCRIPTION
**Pull request by FlyersPh9**
- Add high contrast light theme with AAA score.
- Add high contrast dark theme with AAA score.
- Updated `theme-button`.
	- Updated scenarios of demo pages that included the `theme-button` in the test results since these tests were already visually changing.
- `::selection` opacity changed from opacity 4 to 5.
- Add `color.html` demo page for testing contrast easily.
- Demo.css changes:
	- Use CSS variables for `<hr>`.
	- Add `overflow: overlay` to test scrollbar above different background colors in `color.html`.
- Introducing `$iui-color-background-border` which displays background-5 in default themes but shows foreground-body w/ opacity in high contrast themes.
- Tooltip has a white border to help stand out above any background color.
- Selection overlay opacity is lowered.
- Add inner white circle to active carousel dot.
- Muted text over any overlay background color has had opacity removed.  This is to score a AAA in active header button & selected menu item.
- Header secondary text is using smaller font to align with menu items.
- Active menu item uses blue text to align with header, side nav.

### CODE REVIEW

File: backstop/assets/theme.js
Comment (veekeys) [inline backstop/assets/theme.js:1]: Would it make sense to have another button to reach those themes faster...?

**Review by veekeys — COMMENTED** [review]
I really liked dark theme and pink color the most!
Light theme was weird, though... Looked even more authentic than current one

File: backstop/assets/theme.js
Comment (bentleyvk) [reply backstop/assets/theme.js:1]: And all our test images changes 😕 Maybe right-click could expand dropdown with all themes.

Comment (veekeys) [reply backstop/assets/theme.js:1]: Most of our test images should take sections with components, not full page, I think...

Comment (FlyersPh9) [reply backstop/assets/theme.js:1]: For test images that do have the theme button within, we should be using something like `hideSelectors: ['theme-button'],` in the scenario JS.

File: backstop/assets/theme.js
Comment (mayank99) [inline backstop/assets/theme.js:40]: This will not work in Firefox and Safari 🤔 Could we just use `currentColor` instead?

File: src/style/theme.scss
Comment (veekeys) [reply src/style/theme.scss:78]: You can read `hsl` colors? :O

Comment (mayank99) [reply src/style/theme.scss:78]: Honestly, it is much more intuitive and easy to understand compared to hex or rgb. 

Can you read hex colors? What color is `#965608` and how would you make it 20% lighter?

File: backstop/assets/theme.js
Comment (mayank99) [reply backstop/assets/theme.js:1]: see #511

File: backstop/assets/theme.js
Comment (mayank99) [reply backstop/assets/theme.js:40]: fixed in #511

**Comment by mayank99** [comment]
After #508 and #511 get merged, we also need to check two more things:
1. Contrast ratio should be at least 7:1 to meet WCAG AAA requirements (otherwise it's not really "high" contrast). 4.5:1 should be met in our normal themes.
   - https://www.w3.org/WAI/WCAG21/Understanding/contrast-enhanced.html
   - https://webaim.org/articles/contrast/#sc146
   - https://blog.prototypr.io/high-contrast-mode-ftw-bef74e097d1e
2. Should work with forced high contrast mode in Windows.
   - https://blogs.windows.com/msedgedev/2020/09/17/styling-for-windows-high-contrast-with-new-standards-for-forced-colors/
   - https://blog.khanacademy.org/making-websites-work-with-windows-high-contrast-mode/

**Comment by veekeys** [comment]
> After #508 and #511 get merged, we also need to check two more things:
> 
> 1. Contrast ratio should be at least 7:1 to meet WCAG AAA requirements (otherwise it's not really "high" contrast). 4.5:1 should be met in our normal themes.
>    
>    * [w3.org/WAI/WCAG21/Understanding/contrast-enhanced.html](https://www.w3.org/WAI/WCAG21/Understanding/contrast-enhanced.html)
>    * [webaim.org/articles/contrast/#sc146](https://webaim.org/articles/contrast/#sc146)
>    * [blog.prototypr.io/high-contrast-mode-ftw-bef74e097d1e](https://blog.prototypr.io/high-contrast-mode-ftw-bef74e097d1e)
> 2. Should work with forced high contrast mode in Windows.
>    
>    * [blogs.windows.com/msedgedev/2020/09/17/styling-for-windows-high-contrast-with-new-standards-for-forced-colors](https://blogs.windows.com/msedgedev/2020/09/17/styling-for-windows-high-contrast-with-new-standards-for-forced-colors/)
>    * [blog.khanacademy.org/making-websites-work-with-windows-high-contrast-mode](https://blog.khanacademy.org/making-websites-work-with-windows-high-contrast-mode/)

You can check page situation with Css Overview in Chrome dev tools.

image

Not sure how trustful it is, though...

**Comment by FlyersPh9** [comment]
> 1. Contrast ratio should be at least 7:1 to meet WCAG AAA requirements (otherwise it's not really "high" contrast). 4.5:1 should be met in our normal themes.

Our current light / dark themes do not meet WCAG AA.  The new high contrast light / dark theme meet WCAG AA but not AAA.  We could make these new themes the defaults since they do meet AA and then create new high contrast themes that meet AAA.

File: backstop/tests/color.html
Comment (FlyersPh9) [inline backstop/tests/color.html:249]: Positive / Warning / Negative muted text on overlay background are not present anywhere in our UI.  However primary muted text on overlay is present in the header buttons.

File: backstop/tests/color.html
Comment (FlyersPh9) [inline backstop/tests/color.html:220]: I expanded on the tests on this page and discovered yesterday that muted text on overlay primary is failing in light HC.  This is the only fail I am finding across both light & dark HC themes.  Solutions:
1. Increase `opacity-3` to passing.
2. Change muted text to use `opacity-2`.
3. Remove instances of muted text over primary overlay.

Muted text over primary overlay are present in header buttons, hovered select menu items for examples.

**Comment by FlyersPh9** [comment]
> 1. Contrast ratio should be at least 7:1 to meet WCAG AAA requirements (otherwise it's not really "high" contrast). 4.5:1 should be met in our normal themes.

As of right now all text, with [1 exception](https://github.com/iTwin/iTwinUI/pull/505#discussion_r827270536), is passing with AAA.  Same goes for important interface elements (border of checkboxes) and icons.

> 2. Should work with forced high contrast mode in Windows.

`@media (forced-colors: active)` is new to me.  I will look into if I can replicate this on macOS.  From what I'm seeing [here](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors#css) this might be on a component by component basis.

**Comment by mayank99** [comment]
@FlyersPh9 This looks a lot better! I tried to go through all components, and most places, it's passing AAA, but some of the hover/focus/active states are not.
image
image

Also, disabled/muted/placeholders are obviously failing. I know they are not required to pass AAA, but we might still want to increase the contrast a little bit because some of them are really hard to see.
image

The muted text is also used in footer, which has low visibility and is hard to even tell that it's made of links.
---

One more issue I did notice is that because the dark theme background is so... dark, all our shadows are lost.
image

This is also an issue with translucent backgrounds such as our tooltip where the boundary is lost.
image

So you might want to adjust some of the background colors a little bit, or add borders. https://twitter.com/steveschoger/status/1499780803550658567

File: backstop/tests/color.html
Comment (FlyersPh9) [reply backstop/tests/color.html]: This is addressed in #529.

**Comment by mayank99** [comment]
> [Lines 396-408](https://github.com/iTwin/iTwinUI/blob/2f9b08228690d379a86a64bcb1a1fad432832551/src/style/theme.scss#L396) seem to be overriding changes set above.
> 
> The placeholder text opacity is set to 5, and I'd like to try changing it to 4 for high contrast themes, but I'm not seeing the change.

Looks like it is _meant_ to override the other stuff and it's not possible to add theme-specific values there. You could try moving the `iui-text-color-placeholder` from `$cssVariables` into each of the themes above.

**Review by elephantcatdog — COMMENTED** [review]
I'm not sure how much we control of this, but the colors of many of the user icons don't pass AAA (or even AA). The color on the background or the (white) initials on the color.

Not sure how important this is, but there are four images that don't visually appear to have changed - the dark and light versions of the basic modal and the "session timed out" screen.

**Comment by FlyersPh9** [comment]
> I'm not sure how much we control of this, but the colors of many of the user icons don't pass AAA (or even AA). The color on the background or the (white) initials on the color.

Hmm, I'm not sure how we can best handle this.  Maybe we add a `text-shadow` only when high contrast is on?  Let me think on this for a little bit.

**Comment by elephantcatdog** [comment]
I don't know how well this would work, but you could add an overlay with box-shadow so that user icons are shaded with white for dark theme and black for light theme, and the text color should be black for dark theme (light icons) and white for light theme (dark icons). This does mess with the box-shadows already present though. I don't like the way this looks exactly - something is off with the text color, and I didn't get it working for the stacked lists, but something like this:
image
image
1st one has `box-shadow: inset 0 0 0 200px rgb(255 255 255 / 50%); color: black;` and 2nd one has `box-shadow: inset 0 0 0 200px rgb(0 0 0 / 60%);` in `iui-initials`. But with this method you have to make assumptions about how light/dark the color is initially.

File: src/tooltip/tooltip.scss
Comment (mayank99) [inline src/tooltip/tooltip.scss:23]: Can we avoid this change in default themes? Looks really bad in default dark

File: src/skip-to-content/skip-to-content.scss
Comment (mayank99) [inline src/skip-to-content/skip-to-content.scss:6]: I will say it again, can we avoid this change in default dark theme?

File: src/carousel/carousel.scss
Comment (mayank99) [inline src/carousel/carousel.scss:128]: box-shadow will not work with forced colors mode.

File: src/carousel/carousel.scss
Comment (mayank99) [inline src/carousel/carousel.scss:104]: These changes are going to conflict with #544 so we should decide now how we want to handle that

File: src/style/global.scss
Comment (FlyersPh9) [reply src/style/global.scss:78]: How I about I remove this from this PR and introduce an anchor component _after_ this one merges?

File: src/carousel/carousel.scss
Comment (FlyersPh9) [reply src/carousel/carousel.scss:128]: I've added `@media (forced-colors: active) { border-width: $iui-icons-small * 0.5; }`.
Screen Shot 2022-03-23 at 2 11 50 PM

File: src/tooltip/tooltip.scss
Comment (FlyersPh9) [reply src/tooltip/tooltip.scss:23]: This one I feel pretty strongly that we need to leave in.  The tooltip could appear above an AppUI frame and we have no idea what color might be underneath.
Screen Shot 2022-03-23 at 2 14 21 PM
 
 I wouldn't say it looks bad either, both GitHub and VSC have a light border on their tooltips.

File: src/style/global.scss
Comment (FlyersPh9) [reply src/style/global.scss]: I pushed this change.  I ran tests & tried testing myself to make sure this works and it appears to be fine.

**Review by mayank99 — APPROVED** [review]

**Review by mayank99 — APPROVED** [review]

File: src/style/global.scss
Comment (elephantcatdog) [inline src/style/global.scss:20]: Do you not need the extra & that was previously included with &.iui-theme-light? Looks like it was part of the fallback?

**Review by elephantcatdog — CHANGES_REQUESTED** [review]

File: src/style/global.scss
Comment (FlyersPh9) [reply src/style/global.scss:20]: I believe line 6 makes it so we don't need the extra `&`.

This:
```scss
html {
	@include theme-variables(light);

	&.iui-theme-light {
		@include theme-variables(light);
	}
}
```

Should have the same output as this:
```scss
html {
	&,
	&.iui-theme-light {
		@include theme-variables(light);
	}
}
```

The second option has less text, but I think the first option needs to be placed in this order so it runs through the media queries first & then if the class is applied, it overrides whatever was set by the media query.

**Comment by FlyersPh9** [comment]
> I don't know how well this would work, but you could add an overlay with box-shadow so that user icons are shaded with white for dark theme and black for light theme, and the text color should be black for dark theme (light icons) and white for light theme (dark icons). This does mess with the box-shadows already present though. I don't like the way this looks exactly - something is off with the text color, and I didn't get it working for the stacked lists, but something like this:
1st one has `box-shadow: inset 0 0 0 200px rgb(255 255 255 / 50%); color: black;` and 2nd one has `box-shadow: inset 0 0 0 200px rgb(0 0 0 / 60%);` in `iui-initials`. But with this method you have to make assumptions about how light/dark the color is initially.

That's an interesting approach.  I vote that we let this PR pass as is and tackle both user icons and badge AAA scoring issues in separate PRs while working towards CSS 1.0.

**Review by elephantcatdog — APPROVED** [review]
Looks good to me

File: src/carousel/carousel.scss
Comment (FlyersPh9) [reply src/carousel/carousel.scss:104]: If you are planning to come into the office tomorrow, let's figure it out in person.

**Pull request merged**
